src/UserDirectoryModel.cpp

```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: 2021 Nheko Contributors
+// SPDX-FileCopyrightText: 2022 Nheko Contributors
+// SPDX-FileCopyrightText: 2023 Nheko Contributors
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "UserDirectoryModel.h"
+
+#include "Cache.h"
+#include "Logging.h"
+#include "MatrixClient.h"
+#include "mtx/responses/users.hpp"
+
+UserDirectoryModel::UserDirectoryModel(QObject *parent)
+  : QAbstractListModel{parent}
+{
+    connect(this,
+            &UserDirectoryModel::fetchedSearchResults,
+            this,
+            &UserDirectoryModel::displaySearchResults,
+            Qt::QueuedConnection);
+}
+
+QHash<int, QByteArray>
+UserDirectoryModel::roleNames() const
+{
+    return {
+      {Roles::DisplayName, "displayName"},
+      {Roles::Mxid, "userid"},
+      {Roles::AvatarUrl, "avatarUrl"},
+    };
+}
+
+void
+UserDirectoryModel::setSearchString(const QString &f)
+{
+    userSearchString_ = f.toStdString();
+    nhlog::ui()->debug("Received user directory query: {}", userSearchString_);
+    beginResetModel();
+    results_.clear();
+    endResetModel();
+    searchingUsers_ = true;
+    emit searchingUsersChanged();
+    http::client()->search_user_directory(
+      userSearchString_,
+      [this](const mtx::responses::Users &res, mtx::http::RequestErr err) {
+          searchingUsers_ = false;
+          emit searchingUsersChanged();
+
+          if (err) {
+              nhlog::net()->error("Failed to retrieve users from mtxclient - {} - {} - {}",
+                                  mtx::errors::to_string(err->matrix_error.errcode),
+                                  err->matrix_error.error,
+                                  err->parse_error);
+          } else {
+              emit fetchedSearchResults(res.results);
+          }
```

Comment:
This is actually unsafe. I noticed that only just now, but you might want to do something like this: https://github.com/Nheko-Reborn/nheko/commit/0b8709a0ead4bdc71666f09860a058c90e3c7a1a

Reply:
Not sure what you mean, is this OK?

Reply:
No, you are still not using the proxy as I did in that commit :)

Reply:
I have no idea what I am doing, but does this work?